File: web/packages/teleport/src/services/bot/bot.ts
/**
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { MutationFunction } from '@tanstack/react-query';

import cfg from 'teleport/config';
import api from 'teleport/services/api';
import {
  canUseV1Edit,
  canUseV2Edit,
  makeBot,
  toApiGitHubTokenSpec,
  validateGetBotInstanceMetricsResponse,
  validateGetBotInstanceResponse,
  validateListBotInstancesResponse,
} from 'teleport/services/bot/consts';
import ResourceService, { RoleResource } from 'teleport/services/resources';
import { FeatureFlags } from 'teleport/types';

import { validateListJoinTokensResponse } from '../joinToken/consts';
import { MfaChallengeResponse } from '../mfa';
import { withGenericUnsupportedError } from '../version/unsupported';
import {
  BotResponse,
  CreateBotJoinTokenRequest,
  CreateBotRequest,
  EditBotRequest,
  FlatBot,
} from './types';

export function createBot(
  config: CreateBotRequest,
  mfaResponse?: MfaChallengeResponse
): Promise<void> {
  return api.post(
    cfg.getBotUrl({ action: 'create' }),
    config,
    undefined /* abort signal */,
    mfaResponse
  );
}

export async function getBot(
  variables: {
    botName: string;
  },
  signal?: AbortSignal
): Promise<FlatBot | null> {
  try {
    return await api
      .get(cfg.getBotUrl({ action: 'read', ...variables }), signal)
      .then(makeBot);
  } catch (err) {
    // capture the not found error response and return null instead of throwing
    if (err?.response?.status === 404) {
      return null;
    }
    throw err;
  }
}

export function createBotToken(
  req: CreateBotJoinTokenRequest,
  mfaResponse?: MfaChallengeResponse
) {
  return api.post(
    cfg.getBotTokenUrl(),
    {
      integrationName: req.integrationName,
      joinMethod: req.joinMethod,
      webFlowLabel: req.webFlowLabel,
      gitHub: toApiGitHubTokenSpec(req.gitHub),
    },
    undefined /* abort signal */,
    mfaResponse
  );
}

export async function listBotTokens(
  variables: { botName: string; skipAuthnRetry: boolean },
  signal: AbortSignal
) {
  const path = cfg.getJoinTokenUrl({ action: 'listV2' });
  const qs = new URLSearchParams();
  qs.set('bot_name', variables.botName);
  qs.set('role', 'bot');

  try {
    const data = await api.get(`${path}?${qs.toString()}`, signal, undefined, {
      skipAuthnRetry: variables.skipAuthnRetry,
    });

    if (!validateListJoinTokensResponse(data)) {
      throw new Error('failed to validate list join tokens response');
    }

    return data;
  } catch (err) {
    // TODO(nicholasmarais1158) DELETE IN v20.0.0
    withGenericUnsupportedError(err, '19.0.0');
  }
}

export async function fetchBots(signal: AbortSignal, flags: FeatureFlags) {
  if (!flags.listBots) {
    throw new Error('cannot fetch bots: bots.list permission required');
  }

  return api
    .get(cfg.getBotUrl({ action: 'list' }), signal)
    .then((json: BotResponse) => {
      const items = json?.items || [];
      return { bots: items.map(makeBot) };
    });
}

export async function fetchRoles(
  variables: { search: string; flags: FeatureFlags },
  signal?: AbortSignal
): Promise<{ startKey: string; items: RoleResource[] }> {
  if (!variables.flags.roles) {
    return { startKey: '', items: [] };
  }

  const resourceSvc = new ResourceService();
  return resourceSvc.fetchRoles(
    { limit: 50, search: variables.search },
    signal
  );
}

export const editBotMutationFunction: MutationFunction<
  FlatBot,
  { botName: string; req: EditBotRequest }
> = vars => editBot(vars);

export async function editBot(
  variables: { botName: string; req: EditBotRequest },
  signal?: AbortSignal
) {
  // TODO(nicholasmarais1158) DELETE IN v20.0.0
  const useV1 = canUseV1Edit(variables.req);
  // TODO(nicholasmarais1158) DELETE IN v20.0.0
  const useV2 = canUseV2Edit(variables.req);
  const path = cfg.getBotUrl({
    action: useV1 ? 'update' : useV2 ? 'update-v2' : 'update-v3',
    botName: variables.botName,
  });

  try {
    const res = await api.put(path, variables.req, signal);
    return makeBot(res);
  } catch (err: unknown) {
    // TODO(nicholasmarais1158) DELETE IN v20.0.0
    withGenericUnsupportedError(err, '19.0.0');
  }
}

export async function deleteBot(
  variables: { botName: string },
  signal?: AbortSignal
) {
  return api.deleteWithOptions(
    cfg.getBotUrl({ action: 'delete', botName: variables.botName }),
    {
      signal,
    }
  );
}

export async function listBotInstances(
  variables: {
    pageToken: string;
    pageSize: number;
    searchTerm?: string;
    sortField?: string;
    sortDir?: string;
    botName?: string;
    query?: string;
  },
  signal?: AbortSignal
) {
  const {
    pageToken,
    pageSize,
    searchTerm,
    sortField,
    sortDir,
    botName,
    query,
  } = variables;

  // TODO(nicholasmarais1158) DELETE IN v20.0.0
  const useV1Endpoint = !query;

  const path = cfg.getBotInstanceUrl({
    action: useV1Endpoint ? 'list' : 'listV2',
  });
  const qs = new URLSearchParams();

  qs.set('page_size', pageSize.toFixed());
  qs.set('page_token', pageToken);
  if (searchTerm) {
    qs.set('search', searchTerm);
  }
  if (botName) {
    qs.set('bot_name', botName);
  }

  if (useV1Endpoint) {
    const sort = `${sortField || 'name'}:${sortDir || 'asc'}`;
    if (sort) {
      qs.set('sort', sort);
    }
  } else {
    if (sortField) {
      qs.set('sort_field', sortField);
    }
    if (sortDir) {
      qs.set('sort_dir', sortDir);
    }
    if (query) {
      qs.set('query', query);
    }
  }

  const data = await api.get(`${path}?${qs.toString()}`, signal);

  if (!validateListBotInstancesResponse(data)) {
    throw new Error('failed to validate list bot instances response');
  }

  return data;
}

export async function getBotInstance(
  variables: {
    botName: string;
    instanceId: string;
  },
  signal?: AbortSignal
) {
  const path = cfg.getBotInstanceUrl({ action: 'read', ...variables });

  const data = await api.get(path, signal);

  if (!validateGetBotInstanceResponse(data)) {
    throw new Error('failed to validate get bot instance response');
  }

  return data;
}

export async function getBotInstanceMetrics(
  variables: null,
  signal?: AbortSignal
) {
  const path = cfg.getBotInstanceUrl({ action: 'metrics' });

  try {
    const data = await api.get(path, signal);

    if (!validateGetBotInstanceMetricsResponse(data)) {
      throw new Error('failed to validate get bot instance metrics response');
    }

    return data;
  } catch (err: unknown) {
    // TODO(nicholasmarais1158) DELETE IN v20.0.0
    withGenericUnsupportedError(err, '19.0.0');
  }
}

File: tool/tctl/common/resource/collections/bot.go
/*
 * Teleport
 * Copyright (C) 2025  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package collections

import (
	"io"
	"strings"

	"github.com/gravitational/trace"

	machineidv1pb "github.com/gravitational/teleport/api/gen/proto/go/teleport/machineid/v1"
	"github.com/gravitational/teleport/api/types"
	"github.com/gravitational/teleport/lib/asciitable"
)

func NewBotCollection(bots []*machineidv1pb.Bot) ResourceCollection {
	return &botCollection{bots: bots}
}

type botCollection struct {
	bots []*machineidv1pb.Bot
}

func (c *botCollection) Resources() []types.Resource {
	resources := make([]types.Resource, len(c.bots))
	for i, b := range c.bots {
		resources[i] = types.Resource153ToLegacy(b)
	}
	return resources
}

func (c *botCollection) WriteText(w io.Writer, verbose bool) error {
	t := asciitable.MakeTable([]string{"Name", "Roles"})
	for _, b := range c.bots {
		t.AddRow([]string{
			b.Metadata.Name,
			strings.Join(b.Spec.Roles, ", "),
		})
	}
	_, err := t.AsBuffer().WriteTo(w)
	return trace.Wrap(err)
}

func NewBotInstanceCollection(bots []*machineidv1pb.BotInstance) ResourceCollection {
	return &botInstanceCollection{items: bots}
}

type botInstanceCollection struct {
	items []*machineidv1pb.BotInstance
}

func (c *botInstanceCollection) Resources() []types.Resource {
	r := make([]types.Resource, 0, len(c.items))
	for _, resource := range c.items {
		r = append(r, types.ProtoResource153ToLegacy(resource))
	}
	return r
}

func (c *botInstanceCollection) WriteText(w io.Writer, verbose bool) error {
	headers := []string{"Bot Name", "Instance ID"}

	// TODO: consider adding additional (possibly verbose) fields showing
	// last heartbeat, last auth, etc.
	var rows [][]string
	for _, item := range c.items {
		rows = append(rows, []string{item.Spec.BotName, item.Spec.InstanceId})
	}

	t := asciitable.MakeTable(headers, rows...)

	// stable sort by name.
	t.SortRowsBy([]int{0}, true)
	_, err := t.AsBuffer().WriteTo(w)
	return trace.Wrap(err)
}
